### PR TITLE
Fixing spacer size calculation (revert #18105 changes!)

### PR DESCRIPTION
**Pull request by hamidhadi**
I have a problem with `scrollToLocation` method of `SectionList`. 
See [this discussion](https://github.com/facebook/react-native/issues/22809) or [this repo](https://github.com/hamidhadi/react-native-scroll-spy).
I found that [this pull request](https://github.com/facebook/react-native/pull/18105) changed some calculation on `VirtualizedList`.
So after setting `stickySectionHeadersEnabled` to `false`, the problem was solved.
Finally, I reverted changes of [that pull](https://github.com/facebook/react-native/pull/18105) back and tested again and everything is working now without any problem.

You can see demos of each state to make clear the problem.

⭕️ Current state:

- Sometimes the scroll is jumping while scrolling
- The SectionHeader is jumping while scrolling
- The position of the scroll is not correct when using `scrollToLocation`
- Rendering of rows is not working properly on long lists and some rows push the scroll up or down(according to the direction of scrolling) after rendering.

demo_without_changes

✅ After revert back [#18105](https://github.com/facebook/react-native/pull/18105) changes:

demo_with_changes

## Changelog:
[General][Changed] - Changed calculation of `leadSpace` of sticky section header in `VirtualizedList.js` 

## Test Plan:
Create a app component like below
```javascript
import React, {Component} from 'react';
import {Platform, StyleSheet, Text, View, SectionList, TouchableOpacity} from 'react-native';
import mockData from './MOCK_DATA'
import layoutTableGenerator from './layout_table_generator'

export default class App extends Component {
  state = {
    activeSectionIndex: 0
  }

  generateSection = () => {
    return mockData.reduce((acc, curr) => {
      const sectionIndex = acc.findIndex(section => section.title === curr.color)
      
      if (sectionIndex >= 0) {
        acc[sectionIndex].data.push(curr)
      } else {
        acc.push({ title: curr.color, data: [curr] })
      }
  
      return acc
    }, [])
  }

  renderItem = ({ item, index, section }) => {
    return(
      <View key={item.id} style={{ flex: 1, height: 100, alignItems: 'center', justifyContent: 'center' }}>
        <Text style={{ color: 'red' }}>{index}</Text>
        <Text>{`${item.first_name} ${item.last_name}`}</Text>
        <Text>{`${item.color}`}</Text>
      </View>
    )
  }

  renderSectionHeader = ({section: {title}}) => (
    <View style={{ height: 40, backgroundColor: '#ccc', justifyContent: 'center', borderTopColor: 'red', borderTopWidth: 1 }}>
    <Text style={{fontWeight: 'bold'}}>{title}</Text>
    </View>
  )

  getItemLayout = (
    data,
    index,
  ) => {
    const layoutTable = layoutTableGenerator(data, 100, 40)

    return {length: layoutTable[index].length, offset: layoutTable[index].offset, index}
  }

  keyExtractor = (item, index) => item.id 

  render() {
    const sections = this.generateSection()

    return (
      <View style={styles.container}>
        <SectionList
          renderItem={this.renderItem}
          renderSectionHeader={this.renderSectionHeader}
          sections={sections}
          keyExtractor={this.keyExtractor}
          getItemLayout={this.getItemLayout}
          ref={me => this.sectionList = me}
        />
        <TouchableOpacity
          style={{
            backgroundColor: 'red',
            padding: 5,
            position: 'absolute',
            left: 0,
            bottom: 0
          }}

          onPress={() => {
            if (this.sectionList) {
              const nextIndex = this.state.activeSectionIndex + 3
              if (nextIndex > sections.length) {
                this.setState({ activeSectionIndex: 0 }, () => this.sectionList.scrollToLocation({ sectionIndex: 0, itemIndex: 0, viewPosition: 0 }))
              } else {
                this.setState({
                  activeSectionIndex: nextIndex
                }, () => this.sectionList.scrollToLocation({ sectionIndex: nextIndex, itemIndex: 0, viewPosition: 0, viewOffset: 40 }))
              }
            }
          }}
        >
          <Text style={{ fontWeight: 'bold' }}>Scroll</Text>
        </TouchableOpacity>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#F5FCFF',
  },
  welcome: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10,
  },
  instructions: {
    textAlign: 'center',
    color: '#333333',
    marginBottom: 5,
  },
});
```

Create a `layout_table_generator.js` and paste it to that:
```javascript
module.exports = (sections, itemHeight, sectionHeaderHeight, sectionFooterHeight = 0) => {
  return sections.reduce((layoutTable, section, sectionIndex) => {
    const layoutTableLastItem = layoutTable[layoutTable.length - 1]
    const currentSectionLayoutTable = []
    let CSLTLength

    currentSectionLayoutTable.push({
      length: sectionHeaderHeight,
      offset: layoutTableLastItem ? (layoutTableLastItem.offset + layoutTableLastItem.length) : 0
    })

    for(let i = 0; i < section.data.length; i++) {
      CSLTLength = currentSectionLayoutTable.length

      currentSectionLayoutTable.push({
        length: itemHeight,
        offset: currentSectionLayoutTable[CSLTLength - 1].offset + currentSectionLayoutTable[CSLTLength - 1].length
      })
    }

    CSLTLength = currentSectionLayoutTable.length
    currentSectionLayoutTable.push({
      length: sectionFooterHeight,
      offset: currentSectionLayoutTable[CSLTLength - 1].offset + currentSectionLayoutTable[CSLTLength - 1].length
    })

    return [
      ...layoutTable,
      ...currentSectionLayoutTable
    ]
  }, [])
}
```
And fnally, download [this mock data](https://github.com/hamidhadi/react-native-scroll-spy/blob/master/MOCK_DATA.js) and add it to your project as `MOCK_DATA.js`

Run the project and you can see there is no problem with `scrollToLocation`.

### CODE REVIEW

File: Libraries/Lists/VirtualizedList.js [hunk 1]
```diff
@@ -806,7 +806,7 @@ class VirtualizedList extends React.PureComponent<Props, State> {
             if (stickyIndicesFromProps.has(ii + stickyOffset)) {
               const initBlock = this._getFrameMetricsApprox(lastInitialIndex);
               const stickyBlock = this._getFrameMetricsApprox(ii);
-              const leadSpace = stickyBlock.offset - initBlock.offset;
+              const leadSpace = stickyBlock.offset - (initBlock.offset + initBlock.length);
```
[the remark this answers is not in the document]
Comment (hamidhadi) [reply Libraries/Lists/VirtualizedList.js]: I just committed to fixing this issue.

File: Libraries/Lists/VirtualizedList.js [hunk 2]
```diff
@@ -806,7 +806,9 @@ class VirtualizedList extends React.PureComponent<Props, State> {
             if (stickyIndicesFromProps.has(ii + stickyOffset)) {
               const initBlock = this._getFrameMetricsApprox(lastInitialIndex);
               const stickyBlock = this._getFrameMetricsApprox(ii);
-              const leadSpace = stickyBlock.offset - initBlock.offset;
+              const leadSpace =
+                stickyBlock.offset -
```
[the remark this answers is not in the document]
Comment (hamidhadi) [reply Libraries/Lists/VirtualizedList.js]: @hramos Another issue has just happened! What should I do? 🤦‍♂️

[the remark this answers is not in the document]
Comment (hramos) [reply Libraries/Lists/VirtualizedList.js]: Alright, this is silly. It looks like prettier is not happy either way... You can ignore it, and we'll try to resolve it after importing the PR.

[the remark this answers is not in the document]
Comment (hramos) [reply Libraries/Lists/VirtualizedList.js]: (BTW, did you run prettier locally?)

**Comment by cpojer** [comment]
We recently landed a different PR in this code that was aimed at fixing this issue, see https://github.com/facebook/react-native/pull/22025 so I'm going to close this PR.

**Comment by hamidhadi** [comment]
Hey @cpojer! 
I still have the problem even with that PR that you mentioned. 
What can I do? send another pull request?

**Pull request closed without merging**